File: site-src/examples/algorithm-deepening/pattern-lab-v07/linear_dp_trace.cpp
#include <algorithm>
#include <iostream>
#include <numeric>
#include <tuple>
#include <vector>

struct Result {
  int total;
  std::vector<int> chosen;
  std::vector<int> dp;
};

Result best_non_adjacent(const std::vector<int>& values) {
  std::vector<int> dp(values.size() + 1, 0);
  for (std::size_t prefix = 1; prefix <= values.size(); ++prefix) {
    const int skip = dp[prefix - 1];
    const int take = values[prefix - 1] + (prefix >= 2 ? dp[prefix - 2] : 0);
    dp[prefix] = std::max(skip, take);
  }
  std::vector<int> chosen;
  std::size_t prefix = values.size();
  while (prefix > 0) {
    const int skip = dp[prefix - 1];
    const int take = values[prefix - 1] + (prefix >= 2 ? dp[prefix - 2] : 0);
    if (take > skip) {
      chosen.push_back(static_cast<int>(prefix - 1));
      prefix = prefix >= 2 ? prefix - 2 : 0;
    } else {
      --prefix;
    }
  }
  std::reverse(chosen.begin(), chosen.end());
  return {dp.back(), chosen, dp};
}

std::pair<int, std::vector<int>> highest_first(const std::vector<int>& values) {
  std::vector<int> order(values.size());
  std::iota(order.begin(), order.end(), 0);
  std::sort(order.begin(), order.end(), [&](int left, int right) {
    return std::tie(values[right], left) < std::tie(values[left], right);
  });
  std::vector<bool> blocked(values.size(), false);
  std::vector<int> chosen;
  for (const int index : order) {
    if (values[index] <= 0 || blocked[index]) continue;
    chosen.push_back(index);
    blocked[index] = true;
    if (index > 0) blocked[index - 1] = true;
    if (index + 1 < static_cast<int>(values.size())) blocked[index + 1] = true;
  }
  std::sort(chosen.begin(), chosen.end());
  int total = 0;
  for (const int index : chosen) total += values[index];
  return {total, chosen};
}

void print_values(const std::vector<int>& values) {
  for (std::size_t index = 0; index < values.size(); ++index) {
    if (index > 0) std::cout << ',';
    std::cout << values[index];
  }
}

int main() {
  const std::vector<int> values{4, 5, 4, 1, 1};
  const auto result = best_non_adjacent(values);
  const auto greedy = highest_first(values);
  std::cout << "values=4,5,4,1,1\n";
  std::cout << "dp="; print_values(result.dp); std::cout << '\n';
  std::cout << "optimal=" << result.total << " chosen_indices="; print_values(result.chosen); std::cout << '\n';
  std::cout << "highest_first=" << greedy.first << " chosen_indices="; print_values(greedy.second); std::cout << '\n';
  std::cout << "transition=dp[i]=max(dp[i-1],dp[i-2]+value[i-1])\n";
  std::cout << "tie=skip-current\n";
  std::cout << "invariant=dp-prefix-optimum\n";
}

File: examples/algorithm-deepening/pattern-lab-v07/linear_dp_trace.py
from __future__ import annotations

from itertools import combinations
from typing import Sequence


def best_non_adjacent(values: Sequence[int]) -> tuple[int, tuple[int, ...], tuple[int, ...]]:
    dp = [0] * (len(values) + 1)
    for prefix in range(1, len(values) + 1):
        skip = dp[prefix - 1]
        take = values[prefix - 1] + (dp[prefix - 2] if prefix >= 2 else 0)
        dp[prefix] = max(skip, take)

    chosen: list[int] = []
    prefix = len(values)
    while prefix > 0:
        skip = dp[prefix - 1]
        take = values[prefix - 1] + (dp[prefix - 2] if prefix >= 2 else 0)
        if take > skip:
            chosen.append(prefix - 1)
            prefix -= 2
        else:
            prefix -= 1
    chosen.reverse()
    return dp[-1], tuple(chosen), tuple(dp)


def best_non_adjacent_bruteforce(values: Sequence[int]) -> int:
    best = 0
    indices = range(len(values))
    for size in range(len(values) + 1):
        for chosen in combinations(indices, size):
            if all(chosen[index] - chosen[index - 1] > 1 for index in range(1, len(chosen))):
                best = max(best, sum(values[index] for index in chosen))
    return best


def highest_first(values: Sequence[int]) -> tuple[int, tuple[int, ...]]:
    blocked: set[int] = set()
    chosen: list[int] = []
    for index in sorted(range(len(values)), key=lambda item: (-values[item], item)):
        if values[index] <= 0 or index in blocked:
            continue
        chosen.append(index)
        blocked.update((index - 1, index, index + 1))
    chosen.sort()
    return sum(values[index] for index in chosen), tuple(chosen)


def fixed_report() -> str:
    values = [4, 5, 4, 1, 1]
    total, chosen, dp = best_non_adjacent(values)
    greedy_total, greedy_chosen = highest_first(values)
    return "\n".join(
        [
            "values=4,5,4,1,1",
            f"dp={','.join(map(str, dp))}",
            f"optimal={total} chosen_indices={','.join(map(str, chosen))}",
            f"highest_first={greedy_total} chosen_indices={','.join(map(str, greedy_chosen))}",
            "transition=dp[i]=max(dp[i-1],dp[i-2]+value[i-1])",
            "tie=skip-current",
            "invariant=dp-prefix-optimum",
        ]
    )


if __name__ == "__main__":
    print(fixed_report())
